### ISSUE TITLE
yolo detection sample balance strategy

### ISSUE THREAD

**Issue by OnlySang** [AlexeyAB/darknet#3966]
there are two parameters controlling the sample balance: ignore thresh and truth thresh. ignore thresh defines the negtive sample, and truth thresh defined the positive sample. But, I'm seeing that truth thresh is always 1.0 in all of your cfg files. So it's some experiments' data support this kind of thresh setting? 
I'm trying to realize several sample balance strategy in yolo layer.

**Comment by OnlySang** [comment]
@AlexeyAB In forward_yolo_layer, best iou of each anchor is caculated by prediction. Why not use IOU between anchor and truth directly?

**Comment by AlexeyAB** [comment]
There is compared IoU between anchor and truth directly:
* `pred` is anchor
* `truth_shift` is truth
 https://github.com/AlexeyAB/darknet/blob/05545b260784cb5ce9b27f85462c4a31137b684b/src/yolo_layer.c#L323-L329

**Comment by OnlySang** [comment]
Yes, the code is positive sample; When processing negtive sample, where ignore_thresh take effect, float iou = box_iou(pred, truth);

**Comment by AlexeyAB** [comment]
So why should we do `the same` in both positive and negative parts of code?
It will not duplicate absolutely the same actions?

**Comment by OnlySang** [comment]
It's wired that using prediction of anchor as negtive condition. Maybe it works, but anchor itself not the prediction of anchor that determines whether it is a negative sample or not.
